Apply changes from review.

bryanpaget opened the pull request:
This PR is for a branch containing work from the following branches (merged):

- https://github.com/StatCan/namespace-cleaner/issues/12
- https://github.com/StatCan/namespace-cleaner/issues/13
- https://github.com/StatCan/namespace-cleaner/issues/14
- https://github.com/StatCan/namespace-cleaner/issues/15

### JIRA

- https://jirab.statcan.ca/browse/BTIS-930
- https://jirab.statcan.ca/browse/BTIS-958

Souheil-Yazji commented on cmd/namespace-cleaner/main_test.go:
for ref https://go.dev/blog/godoc

bryanpaget replied:
I have added function-scope comments.

Souheil-Yazji commented on cmd/namespace-cleaner/main.go:
https://go.dev/blog/godoc

bryanpaget replied:
I have added function-scope comments.

Souheil-Yazji commented on tests/job.yaml:
```diff
@@ -1,30 +1,24 @@
-        app: namespace-cleaner
     spec:
-      serviceAccountName: namespace-cleaner
+      serviceAccountName: default
```
see comment about rbac

bryanpaget replied:
I have set the correct serviceAccountName

```yaml
---
apiVersion: batch/v1
kind: Job
metadata:
  name: namespace-cleaner-test-job
  namespace: das
spec:
  template:
    spec:
      serviceAccountName: namespace-cleaner
```

Souheil-Yazji commented on Makefile:
```diff
@@ -1,34 +1,55 @@
+# Integration tests with Kubernetes cluster
+test-integration: docker-build
+	@echo "Running integration tests..."
+	kind load docker-image namespace-cleaner:test
```
is this kind cluster creatable via make?

Souheil-Yazji commented on .github/workflows/test-and-push.yaml:
```diff
@@ -22,39 +20,128 @@ jobs:
+
+      - name: Create Kind cluster with custom config
+        run: |
+          kind create cluster --config - <<EOF
```
possibly extract to makefile

bryanpaget replied:
I removed kind since it was troublesome in github actions and I'm relying on unit testing.

Souheil-Yazji commented on .github/workflows/test-and-push.yaml:
```diff
@@ -22,39 +20,128 @@ jobs:
+          kubectl wait --for=condition=Ready nodes --all --timeout=120s
+          kubectl get nodes -o wide
+
+      - name: Build and load image
```
this is already called by `make test-integration` which is causing the dupe calls

Souheil-Yazji reviewed (changes requested):
really good work on this, we're almost there

bryanpaget commented on .github/workflows/test-and-push.yaml:
```diff
@@ -22,39 +20,128 @@ jobs:
+
+      - name: Run integration tests with enhanced debugging
+        run: |
+          echo "=== Starting integration tests with enhanced debugging ==="
```
I added more logging to validate the final state. I moved most of the kind cluster stuff to the Makefile.

bryanpaget replied:
I have rewritten the github action file so it's much clearer now. I have also made the print statements in the go unit tests conform to a clear template.

bryanpaget commented on .github/workflows/test-and-push.yaml:
```diff
@@ -22,39 +20,128 @@ jobs:
+          echo "Initial namespaces:"
+          kubectl get ns --show-labels
+          echo "Namespace details:"
+          kubectl get ns -o yaml | yq '.items[].metadata | {"name": .name, "labels": .labels, "annotations": .annotations}'
```
Probably not needed. Removed!

bryanpaget replied:
I have removed this.

bryanpaget commented on .github/workflows/test-and-push.yaml:
```diff
@@ -22,39 +20,128 @@ jobs:
+
+          # Run tests with verbose output
+          echo "Running integration tests..."
+          make test-integration
```
I removed duplicate docker builds.

bryanpaget commented:
I have updated the unit tests to address logging issues and I've cleaned up the github action by removing the integration tests.

Summary of changes to unit tests:

- Full GoDoc compliance with function-scope comments
- Clear expected outcomes for each test case
- Consistent documentation style
- Self-explanatory test scenarios
- Easy-to-read test output structure
- Compliance with Go's documentation standards

bryanpaget commented:
## 📦 Overall Scope of the Changes

1. **Complete overhaul of the test suite**

   * Switched from a mix of integration/kind-cluster tests to *pure unit tests*, exercising the core Go logic.
   * Removed the old integration tests, Kind cluster setup, and the Microsoft Graph login bits.
   * Added a broad set of table-driven Go unit tests for:
     – Namespace filtering (valid vs. invalid domains, label parsing)
     – Dry-run vs. delete logic
     – Final-state validation
     – Error and timeout scenarios

2. **Makefile & Dockerfile modernization**

   * Cleaned up and simplified targets for building, testing, linting, and scanning.
   * Switched to multi-stage Alpine images, removed superfluous `chmod +x`, and tidied up `go mod tidy`.
   * Added a dedicated `make test-unit` target, and consolidated CI steps to call that.

3. **GitHub Actions & CI Improvements**

   * Centralized on a single workflow that:

     1. Builds the container
     2. Runs unit tests
     3. Scans with Trivy (exits on critical CVEs)
   * Updated the Buildx invocation, disabled SBOM generation in Docker (per CI speed/size).

4. **Security & RBAC refinements for test-jobs**

   * Added proper ServiceAccount and RoleBinding manifests to the test Job spec.
   * Introduced tolerations on the test Job so it can run on tainted nodes (e.g. control-plane).

5. **Label-time parsing & namespace skipping logic**

   * Switched to a custom `labelTimeLayout` so timestamps on namespace labels parse correctly.
   * Restored “skip invalid-domain” behavior for email-based namespace labels.

6. **Logging & error handling enhancements**

   * Improved debug-level prints throughout `main.go` and the test harness.
   * Made the dry-run flag correctly short-circuit and report simulated deletes.

bryanpaget commented on tests/rbac.yaml:
```diff
@@ -0,0 +1,21 @@
+  name: namespace-cleaner-role
+subjects:
+  - kind: ServiceAccount
+    name: default
```
I set the correct serviceaccounts:

``` yaml
---
apiVersion: batch/v1
kind: Job
metadata:
  name: namespace-cleaner-test-job
  namespace: das
spec:
  template:
    spec:
      serviceAccountName: namespace-cleaner
...
---
apiVersion: rbac.authorization.k8s.io/v1
kind: ClusterRoleBinding
metadata:
  name: namespace-cleaner-bind
subjects:
  - kind: ServiceAccount
    name: namespace-cleaner
    namespace: das
...
```

bryanpaget commented on Makefile:
```diff
@@ -1,63 +1,67 @@
+	@echo "✅ Dry run completed"
+
+# Deployment targets
+run: docker-build ## Deploy to production cluster
```
TODO: Make a follow up to better rename/scope this one.
@bryanpaget
